### DISCUSSION TITLE
优化分页查询API

### DISCUSSION THREAD

**Discussion by LeaFrock** [dotnetcore/FreeSql#1203] [Ideas]
现在分页查询有一些缺陷：

1、`Count(out long count)`因为要支持`out`参数，因此这里实际是同步执行，而非完全的异步。一方面若表行数较大、同步执行的时间较长，会影响接口整体的吞吐量；另一方面使用者需要注意Where条件的位置，若不小心放在Count后可能会出现总数和实际返回结果不一致的情况。

2、在数量查询结果为0时，理论上可以不必执行查询具体数据的SQL，以节省性能；而目前似乎每次都会执行2条SQL。

不过我暂时也提不出具体的改进方案，只是感觉分页查询这里是一个潜在的优化点。

**Comment by 2881099** [comment]
因为异步不支持 out，如果介意异步可以扩展方法解决，也可以单独执行 .CountAsync()

var select = fsql.Select\<T>()..
var count = await select.CountAsync();
var list = await select.ToListAsync();

**Comment by LeaFrock** [comment]
谢谢！建议Wiki文档[这里](https://github.com/dotnetcore/FreeSql/wiki/%e5%88%86%e9%a1%b5%e6%9f%a5%e8%af%a2)补充一下，包括Where条件在Count前后的区别。

**Reply by 2881099** [reply]
文档太复杂了对新人不友好，现在的文档很多人都不仔细看。
只能从使用习惯引导使用者，.Count(out var total) 适合小量数据分页，量大了一般都不会查 count 而采用流式分页，即使 CountAsync 也会很浪费性能。

**Reply by LeaFrock** [reply]
这里也不是很复杂吧 😄 多一点细微处的说明、区分和示例，对充分理解项目的使用以及使用前避坑有很大帮助。Wiki没法提交PR，不然我可以帮忙完善。

> 现在的文档很多人都不仔细看

这种用户就不用管了啊 😆 如果说没有看源码的习惯或者能力还能理解，文档都不看，这么照顾他们作甚。

> .Count(out var total) 适合小量数据分页，量大了一般都不会查 count 而采用流式分页

这些其实都可以放到文档里作为“注解”供使用者参考。类似微软文档里也会插入很多Remarks来辨析具体的使用情形，文档越完善，有助于减少重复或低质提问。

当然，仅仅是建议。

**Reply by 2881099** [reply]
freesql.net 文档，可以在页面最底部点击编辑。

freesql wiki 也是 github 仓库，https://github.com/dotnetcore/FreeSql.wiki.git

**Reply by LeaFrock** [reply]
freesql.net 网站显示无证书，打不开。

freesql wiki是仓库但不能fork和提交PR，直接push应该也不行吧。只能由项目管理者修改。

**Reply by 2881099** [reply]
分页查询文档已加备注
